website/notifications/utils.py

```diff
@@ -0,0 +1,225 @@
+import collections
+from framework.auth.signals import contributor_removed, node_deleted
+from website.models import Node
+from website.notifications.model import Subscription
+from website.notifications.constants import SUBSCRIPTIONS_AVAILABLE, NOTIFICATION_TYPES, USER_SUBSCRIPTIONS_AVAILABLE
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+
+
+class NotificationsDict(dict):
+    def __init__(self):
+        super(NotificationsDict, self).__init__()
+        self.update(messages=[], children=collections.defaultdict(NotificationsDict))
+
+    def add_message(self, keys, messages):
+        """
+        :param keys: ordered list of project ids from parent to node (e.g. ['parent._id', 'node._id'])
+        :param messages: built email message for an event that occurred on the node
+        :return: nested dict with project/component ids as the keys with the message at the appropriate level
+        """
+        d_to_use = self
+        for key in keys:
+            d_to_use = d_to_use['children'][key]
+        if not isinstance(messages, list):
+            messages = [messages]
+        d_to_use['messages'].extend(messages)
+        return True
+
+
+def to_subscription_key(uid, event):
+    return str(uid + '_' + event)
+
+
+def from_subscription_key(key):
+    parsed_key = key.split("_", 1)
+    return {
+        'uid': parsed_key[0],
+        'event': parsed_key[1]
+    }
+
+
+@contributor_removed.connect
+def remove_contributor_from_subscriptions(contributor, node):
+    """ Remove contributor from node subscriptions unless the user is an
+        admin on any of node's parent projects.
+    """
+    if not contributor._id in node.admin_contributor_ids:
```

Comment:
The `Node.remove_contributor` "calls" this function when it sends a `contributor_removed` signal, which is then received by this function. 

Admins with read-permissions can still subscribe to notifications, just like read-only contributor can. 

They are excluded in the case where they are removed from a component as a contributor, but they are an admin on the parent project. This is because the admin still has read-access to the component and thus should still get notifications.